common-lib/src/main/scala/com/gu/mediaservice/GridClient.scala

```diff
@@ -1,123 +1,85 @@
+import play.api.libs.json.{JsArray, JsObject, JsValue, Json, Reads}
 
-import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success, Try}
+import play.api.libs.ws.{WSClient, WSRequest, WSResponse}
 
 object ClientResponse {
   case class Message(errorMessage: String, downstreamErrorMessage: String)
 
   object Message {
-    implicit val reads = Json.reads[Message]
+    implicit val reads: Reads[Message] = Json.reads[Message]
   }
 
   case class DownstreamMessage(errorKey: String, errorMessage: String)
 
   object DownstreamMessage {
-    implicit val reads = Json.reads[DownstreamMessage]
+    implicit val reads: Reads[DownstreamMessage] = Json.reads[DownstreamMessage]
   }
 
   case class Root(message: Message)
 
   object Root {
-    implicit val reads = Json.reads[Root]
+    implicit val reads: Reads[Root] = Json.reads[Root]
   }
 }
 
 case class ClientErrorMessages(errorMessage: String, downstreamErrorMessage: String)
 
-case class ResponseWrapper(body: JsValue, statusCode: Int, bodyAsString: String)
+object GridClient extends LazyLogging {
+  def apply(services: Services)(implicit wsClient: WSClient): GridClient =
+    new GridClient(services)
 
-object GridClient {
-  def apply(apiKey: String, services: Services, maxIdleConnections: Int, debugHttpResponse: Boolean = true): GridClient =
-    new GridClient(apiKey, services, maxIdleConnections, debugHttpResponse)
-}
-
-class GridClient(apiKey: String, services: Services, maxIdleConnections: Int, debugHttpResponse: Boolean) extends ApiKeyAuthentication with LazyLogging {
-
-  import java.util.concurrent.TimeUnit
-
-  private val pool = new ConnectionPool(maxIdleConnections, 5, TimeUnit.MINUTES)
-
-  private val httpClient: OkHttpClient = new OkHttpClient.Builder()
-    .connectTimeout(0, TimeUnit.MINUTES)
-    .readTimeout(0, TimeUnit.MINUTES)
-    .connectionPool(pool)
-    .build()
-
-  def makeGetRequestSync(url: URL, apiKey: String): ResponseWrapper = {
-    val request = new Request.Builder().url(url).header(apiKeyHeaderName, apiKey).build
-    val response = httpClient.newCall(request).execute
-    processResponse(response, url)
+  sealed trait Response {
+    def status: Int
+    def underlying: WSResponse
+    def contentLength: Option[Long] =
+      underlying.header(HeaderNames.CONTENT_LENGTH).flatMap(v => Try(v.toLong).toOption)
   }
 
-  def makeGetRequestAsync(url: URL, apiKey: String)(implicit ec: ExecutionContext): Future[ResponseWrapper] = {
-    makeRequestAsync(url, apiKey).map { response =>
-      processResponse(response, url)
-    }
+  case class Found(json: JsValue, underlying: WSResponse) extends Response {
+    val status = 200
+  }
+  case class NotFound(body: String, underlying: WSResponse) extends Response {
+    val status = 404
   }
+  case class Error(status: Int, url: URL, underlying: WSResponse) extends Response {
+    def logError: Nothing = {
```

Comment:
Given that this is side effecting this should have empty parens at the end to signify that. Also `Unit` is more idiomatic.

Reply:
As it will always throw an exception it needs to be Nothing.  Unit messes with type inference.
Unless you can suggest something better?  I could return the exception and throw it in the calling code?

Reply:
Oh interesting. TIL. Neither of those options are particularly appealing. I try and avoid throwing exceptions with a preference of using an `Either` to communicate back to the caller. But I'm not going to force through another refactor.

As it is, is fine. Empty parens still better and I'd also name the method `logErrorAnd_*cough_sorry*_ThrowException`.